Fix column name crashes

tsibley opened the pull request:
Fixes crashes that occur in `SequenceSheet` loaders (csv, tsv, xls(x), fixed width) when particular column names like `length` or `_fields` or `__name__` are present. Lots of details in the commit messages. Re-fixes #344.

~This PR is based off of #542 so that I could check that no new tests failed; it includes the commits in that PR. A listing of just the commits unique to this PR is here: https://github.com/tsibley/visidata/compare/start-fixing-tests...fix-column-name-crashes.~ _Now independent._

An alternate solution to these crashes could be to not use `namedlist` for user-provided data and remove its use in `SequenceSheet`. It could still be used for hardcoded, VisiData-provided sheets. This seemed liked a bigger change to the internals to me, so I didn't do it, but it is perhaps less fraught with peril in the long run.

tsibley committed:
[tests] Add failing tests demonstrating a tricky bug in namedlist

SequenceSheet rows can not have columns named "length" or "_fields" or
"__name__" due to conflicts with existing attributes on the namedlist
implementation.  This affects formats like csv and tsv but not jsonl,
since the latter does not use SequenceSheet.

The "length" bug was reported in #344 and fixed back in Sept 2019, but
since then the bug was apparently re-introduced.  I did not bisect to
find the re-introduction.

tsibley committed:
[utils] Fix namedlist bug with column named "length"

Columns become attributes on the namedlist instance and a "length"
column overwrote the namedlist.length() method.  This then broke
namedlist.__init__() which calls .length().  Remove .length() and use
the len() builtin instead directly on namedlist._fields.

This re-fixes issue #344 for all SequenceSheets (tsv, csv, xls(x),
fixed_width) and tests/column-name-length.vd now passes.

The more general issue of column names conflicting with the namedlist
implementation remains, and I'll try to address it in the next commit.

tsibley committed:
[utils] Resolve namedlist implementation attrs over column name attrs

Removes explicitly created @propertys for each column.  Instead, column
name attributes will go through the class' __getattr__ and __setattr__
since they don't otherwise exist directly on the instance or class.
These special methods are only called if the attribute doesn't exist, so
implementation-required attributes like _fields and __name__ cannot be
used as column name attributes.  This prevents crashes of vd when such
column names exist in, for example, a csv file.  Columns can still bear
those internal names, however, and be looked up by index on the
namedlist.

The removed chunk of code appears to be from the original namedlist()
implementation introduced 2½ years ago in "Restructure commandlog to
remove movements" (8fb9e6a5).  About a year ago, the __get/setattr__
methods were added in "[vgit] many fixes in status and branches sheets"
(8fc69a61).  I'm not sure why both ways of handling column name
attributes were used, but I believe only one is necessary.

An alternate solution could be to not use namedlist for user-provided
data and remove its use in SequenceSheet.  It could still be used for
hardcoded, vd-provided sheets.  This seemed liked a bigger change to the
internals to me, so I didn't do it, but it is perhaps less fraught with
peril in the long run.

All tests/column-name-*.vd files now pass.  No new test files fail
on this commit compared to before: 15 failures remain.

anjakefala commented:
Hey @tsibley! 

Very nice catch. Thank you so much for looking into this so deeply. 

So far, I have validated the bugs by using your tests. Thank you for including a test for each case. Since each has a unique error, I agree that it was the right thing to do. 

`length` is definitely the most reasonable one to handle. 

In terms of including the test-individually.sh commits I *think* it might make the merge a bit messy. Let's remove them before we press the button. =) I am now going to review the actual change.

anjakefala commented:
Oh! One thing, would you mind moving that `sample_data` to `tests/`?

We like to keep the data that is specific for testing niche use cases in the `tests/` folder. The sample data folder is more for trying out VisiData with various file formats.

anjakefala commented:
@saulpw, for when we all decide whether to go with these changes, or remove namedlist from `SequenceSheet`, here is where `namedlist` is used:

```
class SequenceSheet(Sheet):                                                                                                                                                                  'For sheets which use ColumnItem on rows that are Python sequences (list, namedtuple, etc).'                                                                                        
     def setCols(self, headerrows):                                                                                                                                                               self.columns = []                                                                                                                                                               
         [...]                                                                                                                                                 
         self._rowtype = namedlist('tsvobj', [(c.name or '_') for c in self.columns])                                                                                                    
                                                                                                                                                                                         
     def newRow(self):                                                                                                                                                                   
         return self._rowtype()                                                                                                                                                          
                                                                                                                                                                                         
     def addRow(self, row, index=None):                                                                                                                                                  
             [..]                                                                                                                      
             self._rowtype = namedlist('tsvobj', [(c.name or '_') for c in self.columns])                                                                                                
         if type(row) is not self._rowtype:                                                                                                                                              
             row = self._rowtype(row)                                                                                                                                                      
```

https://github.com/saulpw/visidata/pull/543/commits/d566fa8408040df6675f1e7f729369fc5cf68a87 seems like a very reasonable change to me. I am less comfortable making a call for https://github.com/saulpw/visidata/pull/543/commits/6e250da0ccee44af1a0b2e673c605ca86784187d.

saulpw commented on tests/golden/column-name-_fields.csv:
```diff
@@ -0,0 +1,4 @@
+_fields,foo,bar,baz
```
Is it reasonable to combine these into a single test?  Or do they have to be the first column?

anjakefala replied:
They do not have to be! A single test would be less laborious to scale. The reason why i thought separate tests made sense was bc if it regresses, their individual error messages are not very clear. But it is a lot of files and I really am fine either way.

tsibley replied:
Exactly. The different files cover the different ways the column names can conflict in the implementation. While they could all be put into one file, it would make debugging a regression harder. In fact, I started my local debugging that way, but it wasn't immediately clear when I fixed one way because the others still failed.

saulpw commented on visidata/utils.py:
```diff
@@ -95,8 +91,4 @@ def __setattr__(self, k, v):
             except ValueError:
                 super().__setattr__(k, v)
 
-    for i, attrname in enumerate(fieldnames):
```
Wow, this has been there since 2017.  Thanks for cleaning this up.

tsibley committed:
[tests] Move data for column-name-* tests out of sample_data/

At the request of Anja during review:

    We like to keep the data that is specific for testing niche use
    cases in the tests/ folder. The sample data folder is more for
    trying out VisiData with various file formats.

tsibley commented:
Repushed without the commits from #542 and with data moved from `sample_data/` into `tests/`.

saulpw reviewed (approved):
Thanks @tsibley for fixing this!  I'm sorry if you were burnt by that bug in the first place.

saulpw commented:
By the way, I think it's a great idea to fix TsvSheet/etc to not use namedlist, but I've been putting it off because of the meta-sheets.  Maybe once 2.0 is out..
